README.md
orchard [![Build Status](https://travis-ci.org/kurttheviking/orchard-js.svg?branch=master)](https://travis-ci.org/kurttheviking/orchard-js) [![Coverage Status](https://coveralls.io/repos/github/kurttheviking/orchard-js/badge.svg?branch=master)](https://coveralls.io/github/kurttheviking/orchard-js?branch=master)
=======

Read-through, fail-through, Redis-backed caching


## Getting started

### Prerequisites

- Node.js 4.x (LTS) or newer
- Redis 2.8 or newer

For local development, consider [Docker](https://www.docker.com/) for managing your Redis installation. First install the image:

```sh
docker run -d --name redis -p 6379:6379 redis:2.8
```

Then, run the image:

```sh
docker start redis
```

### Install

Install with [npm](https://www.npmjs.com):

```sh
npm install --save orchard
```

### Use

```js
const Orchard = require('orchard');

const orchard = new Orchard();

orchard(key, () => value).then(console.log);
```


## API

### `Orchard([options])`

This module exports a constructor. Instantiate an Orchard cache with desired `options`:

```js
const Orchard = require('orchard');

const options = {...};

const orchard = new Orchard(options);
```

#### Options

| Option | Type | Description | Default |
| :----- | :--- | :---------- | :------ |
| `herdCacheParams` | `Object` | Parameters passed to the internal, [in-memory LRU cache](https://www.npmjs.com/package/lru-cache) | `{ max: 1024, maxAge: 60000 }` |
| `keyHash` | `Function` | Hashing function applied to resolved keys (prior to prefixing) | An [identity function](https://en.wikipedia.org/wiki/Identity_function) |
| `prefix` | `String` | Prepended to resolved cache keys; use if a single database supports multiple caching services | *None* |
| `scanCount` | `Number` | Hints Redis' `SCAN` command during keyspace enumeration (e.g. with delete) | `10` |
| `ttl` | `String` or `Number` | Sets expiration for a cached value, a `Number` is treated as milliseconds and a `String` is treated as an [`ms` duration](https://www.npmjs.com/package/ms) | *None* |
| `url` | `String` | Redis URI with required authentication | `redis://localhost:6379` |

Notes:

- `herdCacheParams`: To mitigate [thundering herd risks](https://en.wikipedia.org/wiki/Thundering_herd_problem), Orchard holds a `key` in memory while initially resolving the priming value; subsequent calls using the same `key` (within the same process) receive the same `Promise` until the value is resolved. This temporary cache is pruned to avoid unnecessary memory consumption.
- `keyHash`: This function is invoked with the resolved key and should resolve to (or return) a [deterministically hashed digest](https://nodejs.org/api/crypto.html#crypto_crypto_createhash_algorithm) of the input. For example: `function keyHash(key) { return makeDigest(key); }`. By providing a hashing function at instantiation, applications can avoid hashing sensitive keys at every Orchard call-site.
- `ttl`: Redis uses [seconds for expiration timers](https://redis.io/commands/expire); Orchard converts `ttl` from milliseconds to seconds, rounding down.

### `orchard(key, primingFunction[, options])`

Retrieve data from the Orchard cache, invoking the priming function to prime the cache.

```js
const Orchard = requirefastr'orchard');

const orchard = new Orchard(options);

orchard(key, () => Promise.resolve(value)).then(console.log);
```

Returns a `Promise` that resolves to the final value of the `primingFunction`. Internally, Orchard attempts to get the current value of `key` from the database. If the key does not exist or the database is not available, the `primingFunction` is invoked with the resolved `key` (in "raw" form, without the `prefix` and without the `keyHash` function applied) and the final value is set in the database for future use.

`key` can be any one of:

- a `String`
- a `Promise` that resolves to a `String`
- a `Function` that returns a `String`
- a `Function` that returns a `Promise` that resolves to a `String`

`primingFunction` can be any one of:

- a `Function` that returns a [JSON-stringify](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify)-able `Object`
- a `Function` that returns a `Promise` that resolves to a [JSON-stringify](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify)-able `Object`

The `Promise` is rejected if there is an error resolving the `primingFunction`. Rejected cache calls are not persisted.

**Warning:** The `key` resolution process does not have the same stampede protections as the priming function. Instead, if `key` is a `Function`, it is invoked on every cache call. If you plan to remotely resolve the key you may want to consider caching the `key` function as well.

#### Options

| Option | Type | Description | Default |
| :----- | :--- | :---------- | :------ |
| `force` | `Boolean` | If `true`, ignore current cached value and overwrite with new value | `false` |

### `cache#del(key)`

Immediately delete one or more keys and associated data from the Orchard cache.

```js
const Orchard = require('orchard');

const orchard = new Orchard(options);

orchard.del(key).then(console.log);
```

Returns a `Promise` that resolves to the number of removed keys. If the last character of the resolved `key` is an asterisk (`*` &ndash; e.g. `items:*`), the key is treated as a pattern to use with Redis' [`SCAN` and `MATCH` commands](https://redis.io/commands/scan).

### `cache#on(eventName, eventListener)`

Listen to events emitted by Orchard.

```js
cache.on('cache:hit', console.log);
```

The cache instance is also an [event emitter](http://nodejs.org/api/events.html#events_class_events_eventemitter). `eventName` is a `String` corresponding to a supported event and `eventListener` is a callback function invoked each time the event is triggered.

#### `cache:hit`

```js
{
  requestId: <String:UUID>,
  key: <String>,
  ms: <Number:Integer:Milliseconds>
}
```

`ms` is milliseconds elapsed between invocation and resolution of the cached value.

#### `cache:miss`

```js
{
  requestId: <String:UUID>,
  key: <String>,
  ms: <Number:Integer:Milliseconds>
}
```

`ms` is milliseconds elapsed between invocation and resolution of the priming value *before* writing to the database.

#### `redis:error`

```js
<Error>{
  code: <String>,
  errno: <String>,
  syscall: <String>
}
```

#### `redis:ready`

*No event data*


## Development

### Tests

To run the full test suite, including unit tests and linting:

```sh
npm test
```

To run the only the unit test suite:

```sh
npm run test-unit
```

This project maintains 100% coverage of statements, branches, and functions. To determine unit test coverage:

```sh
npm run coverage
```

To run the integrated test suit against a live Redis database (`redis://localhost:6379`):

```sh
npm run test-integrated
```

**Warning:** The integrated test suite deletes all keys prefixed with `__ORCHARD_INTEGRATED_TEST`.

### Debug

This module is instrumented with [debug](https://www.npmjs.com/package/debug) to provide information about behavior during development. All debug invocations use the `orchard` prefix for targeted debug output:

```js
DEBUG=orchard* npm test
```

### Contribute

PRs are welcome! PRs must pass unit tests and linting prior to merge. For bugs, please include a failing test which passes when your PR is applied. To enable a git hook that runs `npm test` prior to pushing, `cd` into your repo and run:

```sh
touch .git/hooks/pre-push
chmod +x .git/hooks/pre-push
echo "npm test" > .git/hooks/pre-push
```

### Versioning

This project follows [semantic versioning](http://semver.org/). See the [change log](CHANGELOG.md) for release information.

### License

- This module is licensed under the [MIT License](./LICENSE)
- Redis is licensed under the ["Three Clause" BSD License](https://redis.io/topics/license)


CHANGELOG.md
## 3.0.0 (Apr 10, 2017)

- [breaking] Priming function receives pre-hash, pre-prefix resolved key

## 2.2.0 (Apr 10, 2017)

- [minor] Add support for key hashing function (`#keyHash`)

## 2.1.0 (Apr 2, 2017)

- [minor] Add support for invocation-level option: force (ignores current value
  )

## 2.0.0 (Apr 1, 2017)

- [breaking] Removed key-level cache options
- [breaking] Removed ability to disable fail-through
- [breaking] Switched to native `Promise` instead of [bluebird](https://www.npmjs.com/package/bluebird)
- [breaking] Switched to ES6 syntax (Node.js 4+ now required)
- [patch] Upgraded dependencies
- [patch] Reorganized test suite


## 1.0.0 (Feb 22, 2016)

- [breaking] Moved key-level options to third argument of the cache call
- [breaking] `#prune` (`#evict`) and `#prunePattern` (`#evictPattern`) consolidated into `#del`
- [breaking] Renamed instance and key-level arguments (e.g. `keyPrefix` &rarr; `prefix`, `expires` &rarr; `ttl`)
- [minor] Generalized key and priming values to any primitive, Promise, or Function
- [patch] Addressed thundering herd risk
- [patch] Upgraded dependencies
- [patch] Reorganized and expanded test suite


lib/Orchard.js
'use strict';

const debug = require('debug')('orchard');
const EventEmitter = require('events').EventEmitter;
const ms = require('ms');
const util = require('util');
const uuid = require('uuid-with-v6');

const parseTtl = require('./utils/parseTtl');
const Redis = require('./services/Redis');
const resolveKey = require('./utils/resolveKey');
const resolveValueFactory = require('./utils/resolveValue');

function Orchard(config) {
  if (!(this instanceof Orchard)) {
    return new Orchard(config);
  }

  const configuration = config || {};
  const self = this;

  EventEmitter.call(self);

  debug('init with config', configuration);

  const herdCacheParams = configuration.herdCacheParams || { max: 1024, maxAge: ms('1m') };
  const keyHash = configuration.keyHash || function keyHashDefault(raw) { return raw; };
  const prefix = configuration.prefix ? String(configuration.prefix) : '';
  const scanCount = configuration.scanCount ? Math.abs(parseInt(configuration.scanCount, 10)) : 10;
  const ttl = parseTtl(configuration.ttl);
  const url = configuration.url || 'redis://localhost:6379';

  let isConnected = false;

  const resolveValue = resolveValueFactory(herdCacheParams);
  const redis = new Redis(url);

  redis.on('ready', () => {
    isConnected = true;

    self.emit('redis:ready');
  });

  redis.on('error', (err) => {
    isConnected = false;

    self.emit('redis:error', err);
  });

  function scanAndDelete(cursor, removedCt, iterationCommands) {
    // [KE] make the current cursor the first element of the executed command array
    //      without modifying the original command array
    const cmds = [cursor].concat(iterationCommands.slice(0));

    return redis.scan(cmds).then((result) => {
      const cursorNew = parseInt(result[0], 10);
      const matchedKeys = result[1];

      return Promise.all(matchedKeys.map(k => redis.del(k))).then((counts) => {
        const updatedRemovedCt = counts.reduce((out, count) => out + count, removedCt);

        // [KE] cursor=0 after fully enumerating the keyspace
        if (cursorNew === 0) {
          return updatedRemovedCt;
        }

        return scanAndDelete(cursorNew, updatedRemovedCt, iterationCommands);
      });
    });
  }

  function cache(key, value, opts) {
    const options = opts || {};
    const requestId = uuid.v6();
    const requestTimestamp = Date.now();

    const force = Boolean(options.force);

    if (key === undefined) {
      return Promise.reject(new Error('missing key'));
    } else if (value === undefined) {
      return Promise.reject(new Error('missing value'));
    }

    debug('[%s] request (connected=%s)', requestId, isConnected);

    return resolveKey(key)
    .then(resolvedKey => Promise.all([resolvedKey, keyHash(resolvedKey)]))
    .then((keyResult) => {
      const rawKey = keyResult[0];
      const hashedKey = keyResult[1];

      const finalKey = `${prefix || ''}${prefix ? ':' : ''}${hashedKey}`;

      debug('[%s] key: %s (%s)', requestId, finalKey, rawKey);

      return [rawKey, finalKey];
    })
    .then((keyResult) => {
      const rawKey = keyResult[0];
      const finalKey = keyResult[1];

      return Promise.all([
        rawKey,
        finalKey,
        (isConnected && !force) ? redis.get(finalKey) : null]
      );
    })
    .then((initialResult) => {
      const rawKey = initialResult[0];
      const finalKey = initialResult[1];
      const cachedValue = initialResult[2];

      if (cachedValue) {
        const finalValue = JSON.parse(cachedValue);

        debug('[%s] cached value', requestId, finalValue);

        self.emit('cache:hit', { requestId, key: finalKey, ms: Date.now() - requestTimestamp });

        return finalValue;
      }

      return resolveValue(rawKey, value).then((primedValue) => {
        debug('[%s] primed value', requestId, primedValue);

        self.emit('cache:miss', { requestId, key: finalKey, ms: Date.now() - requestTimestamp });

        if (!isConnected) {
          return primedValue;
        } else if (ttl) {
          return redis.setex(finalKey, ttl, JSON.stringify(primedValue)).then(() => primedValue);
        }

        return redis.set(finalKey, JSON.stringify(primedValue)).then(() => primedValue);
      });
    });
  }

  cache.del = function pruneKeyspace(key) {
    return resolveKey(key)
    .then((resolvedKey) => {
      const finalKey = `${prefix || ''}${prefix ? ':' : ''}${resolvedKey}`;

      debug('delete: %s', finalKey);

      return finalKey;
    })
    .then((finalKey) => {
      if (finalKey.endsWith('*')) {
        return scanAndDelete(0, 0, ['MATCH', finalKey, 'COUNT', scanCount]);
      }

      return redis.del(finalKey);
    });
  };

  cache.on = function on() {
    self.on.apply(self, arguments);
  };

  return cache;
}

util.inherits(Orchard, EventEmitter);

module.exports = Orchard;
